File: src/main/java/org/trp/shincolle/server/DeskInteractionService.java
package org.trp.shincolle.server;

import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import org.trp.shincolle.block.entity.DeskBlockEntity;
import org.trp.shincolle.entity.base.EntityShipBase;
import org.trp.shincolle.init.ModDataComponents;
import org.trp.shincolle.item.DeskItemBook;
import org.trp.shincolle.menu.DeskMenu;
import org.trp.shincolle.utility.FormationHelper;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.UUID;

public final class DeskInteractionService {
    private DeskInteractionService() {
    }

    public static void updateBookState(Player player, int chapter, int page) {
        if (player == null) {
            return;
        }
        if (player.containerMenu instanceof DeskMenu deskMenu) {
            DeskBlockEntity blockEntity = deskMenu.getBlockEntity();
            if (deskMenu.getDeskType() == 0 && blockEntity != null) {
                blockEntity.setBookChap(chapter);
                blockEntity.setBookPage(page);
                return;
            }
        }

        ItemStack stack = player.getMainHandItem();
        if (!(stack.getItem() instanceof DeskItemBook)) {
            stack = player.getOffhandItem();
        }

        if (stack.getItem() instanceof DeskItemBook) {
            stack.set(ModDataComponents.BOOK_CHAPTER, chapter);
            stack.set(ModDataComponents.BOOK_PAGE, page);
        }
    }

    public static void updateDeskGui(Player player, int guiFunc, int radarZoom) {
        if (player == null) {
            return;
        }
        if (!(player.containerMenu instanceof DeskMenu deskMenu)) {
            return;
        }

        DeskBlockEntity blockEntity = deskMenu.getBlockEntity();
        if (deskMenu.getDeskType() == 0 && blockEntity != null) {
            blockEntity.setGuiFunc(guiFunc);
            blockEntity.setRadarZoomLv(radarZoom);
            if (guiFunc >= 3 && guiFunc <= 4 && player instanceof ServerPlayer serverPlayer) {
                TeamDiplomacyService.sendDeskDiplomacySync(serverPlayer);
            }
        }
    }

    public static void openOwnedShipFromDesk(Player player, UUID shipUuid) {
        if (player == null) {
            return;
        }
        if (!(player.level() instanceof ServerLevel serverLevel)) {
            return;
        }

        Entity entity = serverLevel.getEntity(shipUuid);
        if (entity instanceof EntityShipBase ship && ship.isOwnedBy(player)) {
            ship.openShipMenu(player);
        }
    }

    public static void summonOwnedShipsToDesk(Player player, Collection<UUID> shipUuids) {
        if (player == null
                || !(player.level() instanceof ServerLevel serverLevel)
                || !(player.containerMenu instanceof DeskMenu deskMenu)
                || deskMenu.getDeskType() != 0
                || deskMenu.getBlockEntity() == null
                || shipUuids == null
                || shipUuids.isEmpty()) {
            return;
        }

        ArrayList<UUID> ownedShips = new ArrayList<>();
        HashSet<UUID> seen = new HashSet<>();
        for (UUID shipUuid : shipUuids) {
            if (shipUuid == null || !seen.add(shipUuid)) {
                continue;
            }
            Entity entity = serverLevel.getEntity(shipUuid);
            if (entity instanceof EntityShipBase ship && ship.isOwnedBy(player) && ship.isAlive() && !ship.isInDeadPose()) {
                ownedShips.add(shipUuid);
            }
        }

        if (!ownedShips.isEmpty()) {
            FormationHelper.applySummonShipsToDesk(player, deskMenu.getBlockEntity().getBlockPos(), ownedShips);
        }
    }
}

File: src/main/java/org/trp/shincolle/server/DeskInteractionService.kt
package org.trp.shincolle.server

import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.player.Player
import org.trp.shincolle.block.entity.DeskBlockEntity
import org.trp.shincolle.entity.base.EntityShipBase
import org.trp.shincolle.init.ModDataComponents
import org.trp.shincolle.item.DeskItemBook
import org.trp.shincolle.menu.DeskMenu
import org.trp.shincolle.utility.FormationHelper
import java.util.*

object DeskInteractionService {
    @JvmStatic
    fun updateBookState(player: Player?, chapter: Int, page: Int) {
        if (player == null) {
            return
        }
        val containerMenu = player.containerMenu
        if (containerMenu is DeskMenu) {
            val blockEntity: DeskBlockEntity? = containerMenu.blockEntity
            if (containerMenu.deskType == 0 && blockEntity != null) {
                blockEntity.setBookChap(chapter)
                blockEntity.setBookPage(page)
                return
            }
        }

        var stack = player.getMainHandItem()
        if (stack.getItem() !is DeskItemBook) {
            stack = player.getOffhandItem()
        }

        if (stack.getItem() is DeskItemBook) {
            stack.set<Int?>(ModDataComponents.BOOK_CHAPTER, chapter)
            stack.set<Int?>(ModDataComponents.BOOK_PAGE, page)
        }
    }

    @JvmStatic
    fun updateDeskGui(player: Player?, guiFunc: Int, radarZoom: Int) {
        if (player == null) {
            return
        }
        val containerMenu = player.containerMenu
        if (containerMenu !is DeskMenu) {
            return
        }

        val blockEntity: DeskBlockEntity? = containerMenu.blockEntity
        if (containerMenu.deskType == 0 && blockEntity != null) {
            blockEntity.setGuiFunc(guiFunc)
            blockEntity.setRadarZoomLv(radarZoom)
            if (guiFunc >= 3 && guiFunc <= 4 && player is ServerPlayer) {
                TeamDiplomacyService.sendDeskDiplomacySync(player)
            }
        }
    }

    @JvmStatic
    fun openOwnedShipFromDesk(player: Player?, shipUuid: UUID) {
        if (player == null) {
            return
        }
        if (player.level() !is ServerLevel) {
            return
        }

        val serverLevel = player.level() as ServerLevel
        val entity: Entity? = serverLevel.getEntity(shipUuid)
        if (entity is EntityShipBase
            && entity.isOwnedBy(player)
            && entity.isAlive
            && !entity.isRemoved
        ) {
            entity.openShipMenu(player)
        }
    }

    fun summonOwnedShipsToDesk(player: Player?, shipUuids: MutableCollection<UUID?>?) {
        if (player == null || (player.level() !is ServerLevel)) {
            return
        }
        val containerMenu = player.containerMenu
        if (containerMenu !is DeskMenu || containerMenu.deskType != 0 || containerMenu.blockEntity == null || shipUuids == null || shipUuids.isEmpty()) {
            return
        }

        val serverLevel = player.level() as ServerLevel
        val ownedShips = ArrayList<UUID>()
        val seen = HashSet<UUID>()
        for (shipUuid in shipUuids) {
            if (shipUuid == null || !seen.add(shipUuid)) {
                continue
            }
            val entity: Entity? = serverLevel.getEntity(shipUuid)
            if (entity is EntityShipBase && entity.isOwnedBy(player) && entity.isAlive && !entity.isRemoved && !entity.isInDeadPose) {
                ownedShips.add(shipUuid)
            }
        }

        if (!ownedShips.isEmpty()) {
            FormationHelper.applySummonShipsToDesk(player, containerMenu.blockEntity!!.getBlockPos(), ownedShips)
        }
    }
}
